src/ipld/cid_hashset.rs

```diff
@@ -17,23 +17,139 @@ use cid::Cid;
 // length=32. Taking advantage of this knowledge, we can store the vast majority of CIDs (+99.99%)
 // in one third of the usual space (32 bytes vs 96 bytes).
 #[derive(Default)]
-pub struct CidHashSet {
-    v1_dagcbor_blake2b: HashSet<[u8; BLAKE2B256_SIZE]>,
-    fallback: HashSet<Cid>,
-}
+pub struct CidHashSet(CidHashMap<()>);
 
 impl CidHashSet {
+    /// Adds a value to the set if not already present and returns whether the value was newly inserted.
     pub fn insert(&mut self, cid: Cid) -> bool {
-        match cid.try_into() {
-            Ok(CidVariant::V1DagCborBlake2b(bytes)) => self.v1_dagcbor_blake2b.insert(bytes),
-            Err(()) => self.fallback.insert(cid),
+        if self.0.contains_key(cid) {
+            false
+        } else {
+            self.0.insert(cid, ()).is_none()
         }
```

Comment:
This code should be rewritten. See if you can reason why. If not, ask in the slack channel.